primap2/csg/_strategies/global_least_squares.py

```diff
@@ -0,0 +1,267 @@
+            String representation of fill_ts. Human-readable short representation of
+            the fill_ts (e.g. the source).
+
+        Returns
+        -------
+            filled_ts, descriptions. filled_ts contains the result, where missing
+            data in ts is (partly) filled using scaled data from fill_ts.
+            descriptions contains information about which years were affected and
+            filled how.
+        """
+
+        filled_mask = ts.isnull() & ~fill_ts.isnull()
+        time_filled = filled_mask["time"][filled_mask].to_numpy()
+
+        if time_filled.any():
+            # check of we have overlap. if not use substitution strategy
+            # this might not be necessary because we initialize the LS algorithm with 1,
+            # but better make it explicit
+            overlap = ts.notnull() & fill_ts.notnull()
+            if overlap.any():
+                if self.allow_shift:
+                    #  pass on to lstsq based strategy
+                    strategy = primap2.csg.GlobalLSlstsqStrategy()
+                    filled_ts, descriptions = strategy.fill(
+                        ts=ts,
+                        fill_ts=fill_ts,
+                        fill_ts_repr=fill_ts_repr,
+                    )
+                else:
+                    e = fill_ts[overlap.data].data
+                    e_ref = ts[overlap.data].data
+                    a0 = [1]  #  start with 1 as scaling factor
+                    res = least_squares(
+                        self.factor_mult, a0, jac=self.jac, args=(e, e_ref)
+                    )
+
+                    fill_ts_h = fill_ts * res["x"][0]
+                    filled_ts = xr.core.ops.fillna(ts, fill_ts_h, join="exact")
+
+                    descriptions = [
+                        primap2.ProcessingStepDescription(
+                            time=time_filled,
+                            description="filled with least squares matched data from "
+                            f" {fill_ts_repr}. Factor={res['x'][0]:0.3f}",
+                            function=self.type,
+                            source=fill_ts_repr,
+                        )
+                    ]
+            else:
+                strategy = primap2.csg.SubstitutionStrategy()
+                filled_ts, descriptions = strategy.fill(
+                    ts=ts,
+                    fill_ts=fill_ts,
+                    fill_ts_repr=fill_ts_repr,
+                )
+
+        else:
+            # if we don't have anything to fill we don't need to calculate anything
+            filled_ts = ts
+            descriptions = [
+                primap2.ProcessingStepDescription(
+                    time=time_filled,
+                    description=f"no additional data in {fill_ts_repr}",
+                    function=self.type,
+                    source=fill_ts_repr,
+                )
+            ]
+
+        return filled_ts, descriptions
+
+
+class GlobalLSlstsqStrategy:
```

Comment:
I'd convert this into a single function which just gets called by the other strategy. The user should never need to configure either one or the other strategy.

Reply:
Makes sense for the current code. I kept it as an individual strategy because I thought maybe I'll add a faster version for the b=0 case to this strategy.